In [9]:
import os

import pickle

import numpy as np
import pandas as pd

# Dataset 내부 살펴보기!

In [10]:
DATASET_DIR = "datasets/ASSIST2009/"
DATASET_PATH = os.path.join(DATASET_DIR, "skill_builder_data.csv")

In [11]:
DATASET_PATH

'datasets/ASSIST2009/skill_builder_data.csv'

q_seqs.pkl, r_seqs.pkl

- "각 유저마다" 푼 'skill_id' 와 그 문제에 대한 'correct' 정보를 시퀀스로 저장!

q_list.pkl

- skill_name 유니크 값

u_list.pkl

- user_id 유니크 값

q2idx.pkl, u2idx.pkl

- skill_name, user_id 각각 idx 매겨서 dictionary 로 저장

In [12]:
with open(os.path.join(DATASET_DIR, "q_seqs.pkl"), "rb") as f:
    ex_q_seqs = pickle.load(f)
with open(os.path.join(DATASET_DIR, "r_seqs.pkl"), "rb") as f:
    ex_r_seqs = pickle.load(f)
with open(os.path.join(DATASET_DIR, "q_list.pkl"), "rb") as f:
    ex_q_list = pickle.load(f) # 110개
with open(os.path.join(DATASET_DIR, "u_list.pkl"), "rb") as f:
    ex_u_list = pickle.load(f) # 4151 학생
with open(os.path.join(DATASET_DIR, "q2idx.pkl"), "rb") as f:
    ex_q2idx = pickle.load(f)
with open(os.path.join(DATASET_DIR, "u2idx.pkl"), "rb") as f:
    ex_u2idx = pickle.load(f)

In [13]:
print("ex_q_seqs length: {}".format(len(ex_q_seqs)))
print("ex_r_seqs length: {}".format(len(ex_r_seqs)))
print("ex_q_list length: {}".format(len(ex_q_list)))
print("ex_u_list length: {}".format(len(ex_u_list)))
print("ex_q2idx length: {}".format(len(ex_q2idx)))
print("ex_u2idx length: {}".format(len(ex_u2idx)))



ex_q_seqs length: 4151
ex_r_seqs length: 4151
ex_q_list length: 110
ex_u_list length: 4151
ex_q2idx length: 110
ex_u2idx length: 4151


In [14]:
ex_q_seqs[:3]

[array([67, 18, 67, 18, 67, 18, 18, 67, 67, 18, 18, 67, 32, 35, 18, 32, 18,
        74, 32, 18, 32, 18, 74, 32, 18, 32, 35, 18, 49, 77, 77, 49, 77, 49,
        49]),
 array([48, 99, 48, 99, 48, 99, 99, 48, 49, 99, 99, 49, 49, 99, 54, 54, 54,
        54, 54, 54, 54, 54, 54, 54]),
 array([31, 31, 31, 31, 30, 30])]

In [19]:
ex_r_seqs[:3]

[array([0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
        0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0]),
 array([0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1]),
 array([0, 1, 1, 1, 1, 1])]

In [7]:
ex_u_list.shape[0]

4151

In [8]:
# padding!!!!!

from models.utils import match_seq_len

seq_len = 100
q, r = match_seq_len(ex_q_seqs, ex_r_seqs, seq_len) # seq_len 길이에 맞게 array 조정!

In [9]:
len(q) #아 대박 이해했음. seq_len 을 넘어가는 데이터셋은 잘라서 저장하기 때문에 그런거였음!!!

6204

In [10]:
q[:3]

[array([67, 18, 67, 18, 67, 18, 18, 67, 67, 18, 18, 67, 32, 35, 18, 32, 18,
        74, 32, 18, 32, 18, 74, 32, 18, 32, 35, 18, 49, 77, 77, 49, 77, 49,
        49, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]),
 array([48, 99, 48, 99, 48, 99, 99, 48, 49, 99, 99, 49, 49, 99, 54, 54, 54,
        54, 54, 54, 54, 54, 54, 54, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]),
 array([31, 31, 31, 31, 30, 30, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, 

In [11]:
r[:3]

[array([ 0,  0,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  1,  1,  0,  0,  0,  1,  1,  0,  0,  0,  0,  1,  1,  0,  1,  0,
         0, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]),
 array([ 0,  0,  1,  1,  1,  1,  1,  1,  0,  0,  1,  1,  0,  0,  1,  1,  1,
         1,  1,  1,  1,  1,  1,  1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]),
 array([ 0,  1,  1,  1,  1,  1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, 

In [12]:
len(ex_r_seqs)

4151

# 위 pkl 파일 test / assist2009.py 참고!

## user_id(4151명), skill_name(110개), correct 만 사용함!

**한 데이터셋 묶음**은 '한 학생'이 푼 'skill' 들에 대한 '정오답' 여부를 담고 있는거!

In [40]:
# skill_name 결측값 제거, order_id 랑 skill_name 에서 중복 데이터 제거, order_id 기준 오름차순 정렬

df = pd.read_csv(DATASET_PATH).dropna(subset=["skill_name"])\
            .drop_duplicates(subset=["order_id", "skill_name"])\
            .sort_values(by=["order_id"])

/tmp/ipykernel_56908/3479599840.py:3: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(DATASET_PATH).dropna(subset=["skill_name"])\


In [41]:
len(df)

325637

In [42]:
df.head()

order_id  assignment_id  user_id  assistment_id  problem_id  original  \
337244  20224085         232368    73963          42904       76429         0   
337245  20224095         232368    73963          42904       76430         0   
337246  20224113         232368    73963          42904       76431         0   
337247  20224123         232368    73963          42904       76432         0   
337248  20224142         232368    73963          42904       76433         0   

        correct  attempt_count  ms_first_response tutor_mode  ... hint_count  \
337244        0              3             106016      tutor  ...          0   
337245        1              1             194187      tutor  ...          0   
337246        1              1              12734      tutor  ...          0   
337247        1              1             333484      tutor  ...          0   
337248        0              2              52828      tutor  ...          2   

        hint_total  overlap_time  template_id answer_id  answer_text  \
337244           0        112313        31690  128739.0          NaN   
337245           0        194187        31690  128742.0          NaN   
337246           2         12734        31690       NaN           35   
337247           1        333484        31690  128750.0          NaN   
337248           2         65203        31690       NaN          2.5   

        first_action bottom_hint  opportunity  opportunity_original  
337244             0         NaN            1                   NaN  
337245             0         NaN            2                   NaN  
337246             0         NaN            3                   NaN  
337247             0         NaN            4                   NaN  
337248             0         1.0            5                   NaN  

[5 rows x 30 columns]

In [43]:
u_list = np.unique(df["user_id"].values)
q_list = np.unique(df["skill_name"].values)

In [44]:
# 4천명 학생

print(len(np.unique(df["user_id"].values)))
np.unique(df["user_id"].values)

4151


array([   14, 21825, 51950, ..., 96297, 96298, 96299])

In [45]:
# item 개수 110개

print(len(np.unique(df["skill_name"].values)))
np.unique(df["skill_name"].values)

110


array(['Absolute Value', 'Addition Whole Numbers',
       'Addition and Subtraction Fractions',
       'Addition and Subtraction Integers',
       'Addition and Subtraction Positive Decimals',
       'Algebraic Simplification', 'Algebraic Solving',
       'Angles - Obtuse, Acute, and Right',
       'Angles on Parallel Lines Cut by a Transversal', 'Area Circle',
       'Area Irregular Figure', 'Area Parallelogram', 'Area Rectangle',
       'Area Trapezoid', 'Area Triangle', 'Box and Whisker',
       'Calculations with Similar Figures',
       'Choose an Equation from Given Information', 'Circle Graph',
       'Circumference ', 'Complementary and Supplementary Angles',
       'Computation with Real Numbers', 'Congruence',
       'Conversion of Fraction Decimals Percents', 'Counting Methods',
       'D.4.8-understanding-concept-of-probabilities',
       'Distributive Property', 'Divisibility Rules',
       'Division Fractions',
       'Effect of Changing Dimensions of a Shape Prportionall

In [46]:
u2idx = {u: idx for idx, u in enumerate(u_list)}
q2idx = {q: idx for idx, q in enumerate(q_list)}

In [47]:
df_u = df[df["user_id"] == u_list[0]]#.sort_values(by=["order_id"])
df_u

order_id  assignment_id  user_id  assistment_id  problem_id  original  \
172411  21617623         263599       14          53412       93383         1   
3957    21617623         263599       14          53412       93383         1   
172412  21617632         263599       14          53436       93407         1   
3958    21617632         263599       14          53436       93407         1   
172413  21617641         263599       14          53429       93400         1   
3959    21617641         263599       14          53429       93400         1   
3960    21617650         263599       14          53448       93419         1   
172414  21617650         263599       14          53448       93419         1   
172415  21617659         263599       14          53449       93420         1   
3961    21617659         263599       14          53449       93420         1   
3962    21617667         263599       14          53444       93415         1   
172416  21617667         263599       14          53444       93415         1   
127662  21617675         263599       14          53452       93423         1   
211777  21617675         263599       14          53452       93423         1   
3963    21617675         263599       14          53452       93423         1   
127663  21617692         263600       14          36838       57695         1   
3964    21617692         263600       14          36838       57695         1   
219471  21617692         263600       14          36838       57695         1   
127664  21617731         263600       14          37083       58596         1   
3965    21617731         263600       14          37083       58596         1   
127665  21617749         263600       14          36822       57647         1   
3966    21617749         263600       14          36822       57647         1   
219472  21617749         263600       14          36822       57647         1   
127666  21617805         263600       14          37073       58566         1   
3967    21617805         263600       14          37073       58566         1   
127667  21617825         263600       14          37068       58551         1   
211778  21617825         263600       14          37068       58551         1   
3968    21617825         263600       14          37068       58551         1   
47599   21617893         263601       14          52193       91522         1   
55308   21617915         263601       14          52710       92390         1   
55309   21617924         263601       14          52699       92379         1   
47600   21617931         263601       14          51820       91144         1   
55310   21617942         263601       14          52681       92361         1   
47601   21617953         263601       14          51800       91124         1   
47602   21617962         263601       14          51803       91127         1   

        correct  attempt_count  ms_first_response tutor_mode  ... hint_count  \
172411        0              1              26271      tutor  ...          2   
3957          0              1              26271      tutor  ...          2   
172412        1              1              29123      tutor  ...          0   
3958          1              1              29123      tutor  ...          0   
172413        0              1              13779      tutor  ...          2   
3959          0              1              13779      tutor  ...          2   
3960          0              1              16901      tutor  ...          2   
172414        0              1              16901      tutor  ...          2   
172415        0              1              11079      tutor  ...          2   
3961          0              1              11079      tutor  ...          2   
3962          0              1               8244      tutor  ...          2   
172416        0              1               8244      tutor  ...          2   
127662        0              1              21884  

In [48]:
q2idx

{'Absolute Value': 0,
 'Addition Whole Numbers': 1,
 'Addition and Subtraction Fractions': 2,
 'Addition and Subtraction Integers': 3,
 'Addition and Subtraction Positive Decimals': 4,
 'Algebraic Simplification': 5,
 'Algebraic Solving': 6,
 'Angles - Obtuse, Acute, and Right': 7,
 'Angles on Parallel Lines Cut by a Transversal': 8,
 'Area Circle': 9,
 'Area Irregular Figure': 10,
 'Area Parallelogram': 11,
 'Area Rectangle': 12,
 'Area Trapezoid': 13,
 'Area Triangle': 14,
 'Box and Whisker': 15,
 'Calculations with Similar Figures': 16,
 'Choose an Equation from Given Information': 17,
 'Circle Graph': 18,
 'Circumference ': 19,
 'Complementary and Supplementary Angles': 20,
 'Computation with Real Numbers': 21,
 'Congruence': 22,
 'Conversion of Fraction Decimals Percents': 23,
 'Counting Methods': 24,
 'D.4.8-understanding-concept-of-probabilities': 25,
 'Distributive Property': 26,
 'Divisibility Rules': 27,
 'Division Fractions': 28,
 'Effect of Changing Dimensions of a Shape Pr

In [22]:
df_u["skill_name"]

172411              Percent Of
3957              Circle Graph
172412              Percent Of
3958              Circle Graph
172413              Percent Of
3959              Circle Graph
3960              Circle Graph
172414              Percent Of
172415              Percent Of
3961              Circle Graph
3962              Circle Graph
172416              Percent Of
127662    Equivalent Fractions
211777        Finding Percents
3963              Circle Graph
127663    Equivalent Fractions
3964              Circle Graph
219471              Proportion
127664    Equivalent Fractions
3965              Circle Graph
127665    Equivalent Fractions
3966              Circle Graph
219472              Proportion
127666    Equivalent Fractions
3967              Circle Graph
127667    Equivalent Fractions
211778        Finding Percents
3968              Circle Graph
47599                   Median
55308                    Range
55309                    Range
47600                   Median
55310   

In [23]:
q_seq = np.array([q2idx[q] for q in df_u["skill_name"]])
r_seq = df_u["correct"].values

In [24]:
q_seq

array([67, 18, 67, 18, 67, 18, 18, 67, 67, 18, 18, 67, 32, 35, 18, 32, 18,
       74, 32, 18, 32, 18, 74, 32, 18, 32, 35, 18, 49, 77, 77, 49, 77, 49,
       49])

In [25]:
r_seq

array([0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0])

In [27]:
with open(os.path.join(DATASET_DIR, "train_indices.pkl"), "rb") as f:
    ex_train_indices = pickle.load(f)
with open(os.path.join(DATASET_DIR, "test_indices.pkl"), "rb") as f:
    ex_test_indices = pickle.load(f)

In [28]:
print(len(ex_test_indices))

len(ex_train_indices)


621


5583

In [3]:
with open(os.path.join('./ckpts/dkt/ASSIST2009/', "aucs.pkl"), "rb") as f:
    ex_aucs = pickle.load(f)
with open(os.path.join('./ckpts/dkt/ASSIST2009/', "loss_means.pkl"), "rb") as f:
    ex_loss_means = pickle.load(f)

In [4]:
ex_loss_means

[0.6557968,
 0.6218933,
 0.60753334,
 0.6010196,
 0.5959706,
 0.59302384,
 0.59084475,
 0.5882553,
 0.58673155,
 0.5838005,
 0.58321434,
 0.58275795,
 0.58124536,
 0.5794871,
 0.5797563,
 0.5778108,
 0.57589644,
 0.5750578,
 0.5749064,
 0.5737426,
 0.5724571,
 0.57244056,
 0.5722723,
 0.570712,
 0.5700486,
 0.5685785,
 0.56811213,
 0.5676834,
 0.5673427,
 0.5668502,
 0.5652522,
 0.5647806,
 0.5636292,
 0.5635472,
 0.5632827,
 0.56267774,
 0.5611421,
 0.561647,
 0.5619542,
 0.55949366,
 0.55907005,
 0.55886245,
 0.5579101,
 0.5566351,
 0.55712533,
 0.55580825,
 0.55540854,
 0.5542409,
 0.55381566,
 0.55262643,
 0.5524741,
 0.5524965,
 0.55149466,
 0.5514086,
 0.5505556,
 0.5497853,
 0.5485297,
 0.5476242,
 0.54817283,
 0.54603934,
 0.545804,
 0.5453421,
 0.5451136,
 0.5447201,
 0.54352754,
 0.54388165,
 0.5424297,
 0.5426063,
 0.5426936,
 0.54098743,
 0.5394903,
 0.53969055,
 0.5391815,
 0.53897417,
 0.53904176,
 0.5382361,
 0.53757966,
 0.53753066,
 0.53646237,
 0.5347196,
 0.53482825,

In [8]:
max(ex_aucs)

0.818712465873296

# Dataset, DataLoader 살펴보기

### Dataloader 은 batch 기반의 딥러닝 모델 학습을 위해 mini batch 를 만들어주는 역할을 한다.
### 이를 통해 dataset의 전체 데이터가 batch size 로 slice 된다

In [2]:
from data_loaders.assist2009 import ASSIST2009

dataset = ASSIST2009(100)

In [3]:
import torch
from torch.utils.data import DataLoader, random_split

train_size = int(len(dataset) * 0.9)
test_size = len(dataset) - train_size

train_dataset, test_dataset = random_split(
        dataset, [train_size, test_size], generator=torch.Generator(device='cuda')
    )

In [6]:
# train_dataset

train_size #5583, test-621

train_size+test_size #왜???????
#아 대박 이해했음. seq_len 을 넘어가는 데이터셋은 잘라서 저장하기 때문에 그런거였음!!!

# len(dataset)
dataset.__len__() #q_seqs 길이 = 4151 (왜 6204지..?)


6204

In [7]:
train_dataset.indices #전체 데이터셋(6204)에 대해 train에 들어가는 데이터셋의 index

[1629,
 3794,
 4845,
 4597,
 2231,
 5668,
 5931,
 2545,
 3481,
 1650,
 5666,
 3335,
 1725,
 895,
 5627,
 1249,
 1006,
 5962,
 6015,
 5824,
 82,
 4986,
 1418,
 727,
 1759,
 3847,
 6069,
 5765,
 4706,
 955,
 1134,
 1651,
 4658,
 5561,
 1922,
 5863,
 1723,
 2659,
 2091,
 4013,
 6159,
 435,
 1871,
 5132,
 5746,
 1024,
 3710,
 3712,
 1228,
 4996,
 3460,
 1284,
 3524,
 1668,
 3257,
 1399,
 599,
 4552,
 1869,
 5935,
 3983,
 5640,
 1883,
 3554,
 2750,
 3187,
 3599,
 4948,
 4929,
 859,
 4820,
 5626,
 5695,
 5707,
 2982,
 717,
 104,
 4194,
 1659,
 1189,
 1380,
 4933,
 2957,
 2495,
 5332,
 1844,
 1497,
 5289,
 4696,
 1636,
 4462,
 500,
 5156,
 990,
 3692,
 46,
 3792,
 5377,
 298,
 4099,
 888,
 1943,
 6169,
 3102,
 3324,
 3345,
 1681,
 37,
 4322,
 2059,
 1755,
 367,
 2547,
 1705,
 4055,
 2213,
 5349,
 2839,
 4690,
 763,
 559,
 353,
 4052,
 5223,
 3740,
 4335,
 3346,
 2797,
 3293,
 5852,
 441,
 1687,
 5922,
 2954,
 403,
 4514,
 4782,
 1155,
 1245,
 1353,
 1387,
 1361,
 5564,
 1106,
 2795,
 4582,
 1

In [10]:
from models.utils import collate_fn

train_loader = DataLoader(
        train_dataset, batch_size=256, shuffle=True,
        collate_fn=collate_fn, generator=torch.Generator(device='cuda') #0607 hson
    )
test_loader = DataLoader(
        test_dataset, batch_size=test_size, shuffle=True,
        collate_fn=collate_fn, generator=torch.Generator(device='cuda') #0607 hson
    )

In [36]:
# shuffle=False
tmp_loader = DataLoader(
        test_dataset, batch_size=test_size, shuffle=False,
        collate_fn=collate_fn, generator=torch.Generator(device='cuda') #0607 hson
    )

In [40]:
next(iter(tmp_loader))

(tensor([[  1.,   1.,   1.,  ...,  52.,   2.,   2.],
         [ 49.,  67.,  18.,  ...,  -0.,  -0.,  -0.],
         [ 58.,  58.,  59.,  ...,  -0.,  -0.,  -0.],
         ...,
         [  3.,   3.,   3.,  ...,  -0.,  -0.,  -0.],
         [ 67., 102., 102.,  ...,  -0.,  -0.,  -0.],
         [ 23.,  23.,  23.,  ..., 102., 102., 102.]]),
 tensor([[1., 0., 1.,  ..., 1., 0., 1.],
         [0., 0., 0.,  ..., -0., -0., -0.],
         [0., 0., 1.,  ..., -0., -0., -0.],
         ...,
         [0., 1., 1.,  ..., -0., -0., -0.],
         [0., 0., 1.,  ..., -0., -0., -0.],
         [0., 1., 1.,  ..., 0., 0., 1.]]),
 tensor([[  1.,   1.,   1.,  ...,   2.,   2.,   2.],
         [ 67.,  18.,  35.,  ...,  -0.,  -0.,  -0.],
         [ 58.,  59.,  58.,  ...,  -0.,  -0.,  -0.],
         ...,
         [  3.,   3.,   3.,  ...,  -0.,  -0.,  -0.],
         [102., 102., 102.,  ...,  -0.,  -0.,  -0.],
         [ 23.,  23.,  23.,  ..., 102., 102.,  35.]]),
 tensor([[0., 1., 1.,  ..., 0., 1., 1.],
         [0., 0.,

In [9]:
# train_features = iter(train_loader)
# next(train_features)

next(iter(train_loader))

# for data in train_loader:
# #     print(data)
#     q, r, qshft, rshft, m = data
#     print(f"q: {q}\nr: {r}\nqshft: {qshft}\nrshft: {rshft}\nm: {m}")
# #     print(q[0])
# #     print(qshft[0])
#     break

(tensor([[103., 103., 103.,  ...,  -0.,  -0.,  -0.],
         [105., 105., 104.,  ...,  -0.,  -0.,  -0.],
         [ 57.,  90.,  57.,  ..., 106., 106., 106.],
         ...,
         [ 58.,  62.,  62.,  ...,  -0.,  -0.,  -0.],
         [ 52.,  52.,   0.,  ...,  -0.,  -0.,  -0.],
         [  1.,   0.,   3.,  ...,  -0.,  -0.,  -0.]]),
 tensor([[0., 0., 0.,  ..., -0., -0., -0.],
         [0., 1., 1.,  ..., -0., -0., -0.],
         [1., 1., 1.,  ..., 1., 1., 1.],
         ...,
         [1., 0., 0.,  ..., -0., -0., -0.],
         [0., 1., 0.,  ..., -0., -0., -0.],
         [1., 1., 0.,  ..., -0., -0., -0.]]),
 tensor([[103., 103., 103.,  ...,  -0.,  -0.,  -0.],
         [105., 104., 104.,  ...,  -0.,  -0.,  -0.],
         [ 90.,  57.,  90.,  ..., 106., 106., 106.],
         ...,
         [ 62.,  62.,  62.,  ...,  -0.,  -0.,  -0.],
         [ 52.,  52.,  -0.,  ...,  -0.,  -0.,  -0.],
         [  0.,   3.,   3.,  ...,  -0.,  -0.,  -0.]]),
 tensor([[0., 0., 0.,  ..., -0., -0., -0.],
         [1

In [10]:
cnt = 0
for q,r,qshft,rshft,m in test_loader:
    for i in m:
#         print(sum([1 for target in i if target==True]))
#         print(i)
#         break
        cnt += sum([1 for target in i if target==True])
cnt

33149

# model load 해서 inference

In [2]:
import json

with open("config.json") as f:
    config = json.load(f)
    model_config = config["dkt"]
    train_config = config["train_config"]

In [3]:
from data_loaders.assist2009 import ASSIST2009

dataset = ASSIST2009(100)

In [4]:
import torch 
from models.dkt import DKT

model = DKT(dataset.num_q, **model_config)
model.load_state_dict(torch.load('./ckpts/dkt/ASSIST2009/model.ckpt'))


<All keys matched successfully>

In [5]:
model.eval()

DKT(
  (interaction_emb): Embedding(220, 100)
  (lstm_layer): LSTM(100, 100, batch_first=True)
  (out_layer): Linear(in_features=100, out_features=110, bias=True)
  (dropout_layer): Dropout(p=0.5, inplace=False)
)

In [47]:
# dataiter = iter(test_loader)
# q,r,qshft,rshft,m = dataiter.next()

# # print(q,r,qshft,rshft,m)
# print(q)
# q.shape

In [8]:
len(test_dataset.indices)

621

In [11]:
test_loader.batch_size

621

In [53]:
for data in test_loader:
    q, r, qshft, rshft, m = data

In [54]:
data

(tensor([[ 61.,  61.,  61.,  ...,  62.,  62.,  62.],
         [  2.,  20., 100.,  ...,  -0.,  -0.,  -0.],
         [ 62.,  62.,  62.,  ...,  23.,  23.,  23.],
         ...,
         [ 31.,  31.,   0.,  ...,  -0.,  -0.,  -0.],
         [ 73.,  73.,  73.,  ...,  -0.,  -0.,  -0.],
         [  1.,   3.,   3.,  ...,  54.,  54.,  54.]]),
 tensor([[0., 0., 1.,  ..., 1., 1., 1.],
         [1., 1., 0.,  ..., -0., -0., -0.],
         [1., 1., 1.,  ..., 0., 1., 0.],
         ...,
         [0., 0., 0.,  ..., -0., -0., -0.],
         [0., 1., 1.,  ..., -0., -0., -0.],
         [1., 0., 0.,  ..., 0., 0., 1.]]),
 tensor([[ 61.,  61.,  61.,  ...,  62.,  62.,  62.],
         [ 20., 100.,  82.,  ...,  -0.,  -0.,  -0.],
         [ 62.,  62.,  62.,  ...,  23.,  23.,  23.],
         ...,
         [ 31.,  31.,  -0.,  ...,  -0.,  -0.,  -0.],
         [ 73.,  73.,  73.,  ...,  -0.,  -0.,  -0.],
         [  3.,   3.,   3.,  ...,  54.,  54.,  54.]]),
 tensor([[0., 1., 1.,  ..., 1., 1., 1.],
         [1., 0., 0.

In [55]:
print(qshft)
qshft.shape

tensor([[ 61.,  61.,  61.,  ...,  62.,  62.,  62.],
        [ 20., 100.,  82.,  ...,  -0.,  -0.,  -0.],
        [ 62.,  62.,  62.,  ...,  23.,  23.,  23.],
        ...,
        [ 31.,  31.,  -0.,  ...,  -0.,  -0.,  -0.],
        [ 73.,  73.,  73.,  ...,  -0.,  -0.,  -0.],
        [  3.,   3.,   3.,  ...,  54.,  54.,  54.]])


torch.Size([621, 100])

In [56]:
q[2]

tensor([62., 62., 62., 62., 62., 62., 60., 60., 60., 60., 60., 60., 35., 18.,
        32., 67., 18., 32., 35., 18., 35., 18., 32., 18., 67., 32., 35., 18.,
        35., 32., 18., 18., 67., 18., 35., 32., 32., 18., 35., 35., 32., 18.,
        18., 67., 32., 35., 18., 35., 18., 32., 67., 18., 18., 67., 32., 35.,
        18., 18., 35., 32., 18., 32., 35., 18., 67., 23., 23., 23., 23., 23.,
        23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23.,
        23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23.,
        23., 23.])

In [57]:
qshft[2] 

tensor([62., 62., 62., 62., 62., 60., 60., 60., 60., 60., 60., 35., 18., 32.,
        67., 18., 32., 35., 18., 35., 18., 32., 18., 67., 32., 35., 18., 35.,
        32., 18., 18., 67., 18., 35., 32., 32., 18., 35., 35., 32., 18., 18.,
        67., 32., 35., 18., 35., 18., 32., 67., 18., 18., 67., 32., 35., 18.,
        18., 35., 32., 18., 32., 35., 18., 67., 23., 23., 23., 23., 23., 23.,
        23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23.,
        23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23., 23.,
        23., 23.])

In [58]:
r[2]

tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 1., 1., 0.,
        0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 0., 0., 0., 1., 1., 1., 1.,
        0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 0., 0., 1.,
        1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 0.,
        1., 1., 1., 1., 1., 1., 0., 0., 1., 0.])

In [59]:
rshft[2]

tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 1., 1., 0., 0.,
        0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 0., 0., 0., 1., 1., 1., 1., 0.,
        0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 0., 0., 1., 1.,
        0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 0., 1.,
        1., 1., 1., 1., 1., 0., 0., 1., 0., 1.])

In [21]:
from torch.nn.functional import one_hot, binary_cross_entropy
from sklearn import metrics

outputs = model(q.long(), r.long()) #forward 호출 #embedding+LSTM
outputs = (outputs * one_hot(qshft.long(), model.num_q)).sum(-1) #sum(-1): 열의 차원 제거, 행끼리 합 #[621,100]


# masked_select(_, m): padding으로 맞춰줬던 애들 말고 실제 결과에 대해서만 가져오는거임!
# .cpu(): gpu 메모리에 올려져있는 tensor를 cpu 메모리로 복사
outputs = torch.masked_select(outputs, m).detach().cpu() #detach(): 연산 기록으로부터 분리하여 이후 연산 추적 방지
                                                        #이 연산 기록으로부터 도함수가 계산되고 역전파가 이루어짐
t = torch.masked_select(rshft, m).detach().cpu()

In [22]:
outputs

tensor([0.4610, 0.4520, 0.4140,  ..., 0.8970, 0.6809, 0.9480], device='cpu')

In [23]:
t

tensor([0., 1., 1.,  ..., 1., 1., 1.], device='cpu')

In [24]:
t.shape # 33149는 뭐지....?? -> 아 대박 실제 전체 결과에 대한거!!!!!!!!!!!!!!!!!!!!!!!! 


torch.Size([33149])

# 그러니까, 

seq_len 으로 길이를 전부 맞춰줬잖아(100으로)
그리고 최종 연산 결과에서 실제 존재하는 결과에 대해서(skill,correct) 만 뽑아오는거임. padding 으로 채운 값은 제거하고(의미가 없으니까(?))

### 위에 확인해보면 df 전체 325,637개 행 중에서 테스트 데이터셋으로 들어간 행은 33,149개 인거임!!!

In [25]:
auc = metrics.roc_auc_score(
                        y_true=t.numpy(), y_score=outputs.numpy()
                    )

# loss_mean = np.mean(loss_mean)

In [26]:
auc

0.818712465873296

### 여기서 outputs 이, test dataset 에서, 다음 step 에서 skill 에 대한 정오답 확률을 나타내고 있는거 맞나?

그럼 음.. 이제 각 학생을 구분해야 하는거 아닌가?

In [62]:
idx = 101

In [63]:
q[idx]

tensor([77., 95., 48., 95., 95., 77., 48., 95., 77., 95., 34., 46., 46.,  0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0.])

In [64]:
qshft[idx]

tensor([95., 48., 95., 95., 77., 48., 95., 77., 95., 34., 46., 46.,  9., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0.])

In [65]:
r[idx]

tensor([0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.])

In [66]:
rshft[idx]

tensor([0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.,
        -0., -0., -0., -0., -0., -0., -0., -0., -0., -0.])

In [67]:
m[idx]

tensor([ True,  True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False])

In [68]:
from torch.nn.functional import one_hot, binary_cross_entropy
from sklearn import metrics

idx = 101
outputs = model(q[idx].long(), r[idx].long()) #forward 호출 #embedding+LSTM
outputs = (outputs * one_hot(qshft[idx].long(), model.num_q)).sum(-1) #sum(-1): 열의 차원 제거, 행끼리 합 #[621,100]


# masked_select(_, m): padding으로 맞춰줬던 애들 말고 실제 결과에 대해서만 가져오는거임!
# .cpu(): gpu 메모리에 올려져있는 tensor를 cpu 메모리로 복사
outputs = torch.masked_select(outputs, m[idx]).detach().cpu() #detach(): 연산 기록으로부터 분리하여 이후 연산 추적 방지
                                                            #이 연산 기록으로부터 도함수가 계산되고 역전파가 이루어짐
t = torch.masked_select(rshft[idx], m[idx]).detach().cpu()

In [69]:
idx_q = torch.masked_select(qshft[idx], m[idx]).detach().cpu()#.unique()
idx_q = idx_q.tolist()
idx_q

[95.0, 48.0, 95.0, 95.0, 77.0, 48.0, 95.0, 77.0, 95.0, 34.0, 46.0, 46.0, 9.0]

In [70]:
outputs # 다음 번 문제 풀 때 모든 문제에 대해서 이 학생이 각 문제를 맞출 확률 (?) #110개 유형별

tensor([0.1419, 0.2851, 0.1457, 0.4630, 0.8592, 0.4275, 0.8816, 0.7899, 0.8252,
        0.6597, 0.5253, 0.7045, 0.5725], device='cpu')

In [71]:
t

tensor([0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.], device='cpu')

똑같은 99라도,
유형은 같지만 똑같은 문제가 아님

위치로 판별?

LSTM - 시퀀스 순서가 중요
+
순서에 따라 확률이 달라질 수 있음

**평균**? 맥스값?

최종이라는 개념 없이, **조합**, 시퀀스가 중요

한꺼번에 내뱉을 때, ex) 합격을 할/안할 확률

In [72]:
metrics.roc_auc_score(
                        y_true=t.numpy(), y_score=outputs.numpy()
                    )

1.0

In [73]:
unique_q = set(idx_q)
unique_q

{9.0, 34.0, 46.0, 48.0, 77.0, 95.0}

In [75]:
from collections import defaultdict

r_dict = defaultdict(list)
for idx, corr in enumerate(t.tolist()):
    r_dict[idx_q[idx]].append(corr)
r_dict

defaultdict(list,
            {95.0: [0.0, 0.0, 1.0, 1.0, 1.0],
             48.0: [0.0, 1.0],
             77.0: [1.0, 1.0],
             34.0: [1.0],
             46.0: [1.0, 1.0],
             9.0: [1.0]})

In [81]:
q_dict = defaultdict(list)
for idx, p in enumerate(outputs.tolist()):
    q_dict[idx_q[idx]].append(p)
q_dict

q_last_dict = {}
for k,v in q_dict.items():
    q_last_dict[k] = v[-1]
q_last_dict

{95.0: 0.8251921534538269,
 48.0: 0.4274957776069641,
 77.0: 0.7898709774017334,
 34.0: 0.6597404479980469,
 46.0: 0.7045409679412842,
 9.0: 0.5725192427635193}

In [83]:
q_dict

defaultdict(list,
            {95.0: [0.1418902575969696,
              0.14572927355766296,
              0.46296238899230957,
              0.8815793991088867,
              0.8251921534538269],
             48.0: [0.28510886430740356, 0.4274957776069641],
             77.0: [0.8591896295547485, 0.7898709774017334],
             34.0: [0.6597404479980469],
             46.0: [0.5253445506095886, 0.7045409679412842],
             9.0: [0.5725192427635193]})

In [82]:
q_mean_dict = {}
for k,v in q_dict.items():
    q_mean_dict[k] = np.mean(v)
q_mean_dict

{95.0: 0.49147069454193115,
 48.0: 0.35630232095718384,
 77.0: 0.824530303478241,
 34.0: 0.6597404479980469,
 46.0: 0.6149427592754364,
 9.0: 0.5725192427635193}

## 음 근데.. 한 skill 에 대해서 합쳐져서 나오는게 아니라 모든 문제에 대해서 나오네..?

# -> 내가 추가 조정해야 하나 봄

In [49]:
idx2q = {v:k for k,v in q2idx.items()}

In [50]:
idx2q

{0: 'Absolute Value',
 1: 'Addition Whole Numbers',
 2: 'Addition and Subtraction Fractions',
 3: 'Addition and Subtraction Integers',
 4: 'Addition and Subtraction Positive Decimals',
 5: 'Algebraic Simplification',
 6: 'Algebraic Solving',
 7: 'Angles - Obtuse, Acute, and Right',
 8: 'Angles on Parallel Lines Cut by a Transversal',
 9: 'Area Circle',
 10: 'Area Irregular Figure',
 11: 'Area Parallelogram',
 12: 'Area Rectangle',
 13: 'Area Trapezoid',
 14: 'Area Triangle',
 15: 'Box and Whisker',
 16: 'Calculations with Similar Figures',
 17: 'Choose an Equation from Given Information',
 18: 'Circle Graph',
 19: 'Circumference ',
 20: 'Complementary and Supplementary Angles',
 21: 'Computation with Real Numbers',
 22: 'Congruence',
 23: 'Conversion of Fraction Decimals Percents',
 24: 'Counting Methods',
 25: 'D.4.8-understanding-concept-of-probabilities',
 26: 'Distributive Property',
 27: 'Divisibility Rules',
 28: 'Division Fractions',
 29: 'Effect of Changing Dimensions of a Shap

In [178]:
q[100].unique()
# q[100]

tensor([-0.,  3., 15., 48., 49., 51., 74., 77., 87., 95., 96., 99.])

In [183]:
# print(idx2q.get(48))
# print(idx2q.get(49))
# print(idx2q.get(103))

lst = qshft[101].unique()
for i in lst:
    print(int(i), idx2q.get(int(i)))

0 Absolute Value
9 Area Circle
34 Exponents
46 Least Common Multiple
48 Mean
77 Range
95 Stem and Leaf Plot


절댓값, 정수는 중1-1
평균,중앙값,최빈값은 중3-2
줄기와 잎그림은 중1-2

- 참고
    https://m.blog.naver.com/guitarpsy/60205245942

# dkt Training 과정 분석

**collate_fn 살펴보기**

for data in train_loader:

    q, r, qshft, rshft, m = data
                
**부분**

In [118]:
tmp_loader.batch_size
# tmp_loader

print(len(q_seq))
q_seq
# r_seq

35


array([67, 18, 67, 18, 67, 18, 18, 67, 67, 18, 18, 67, 32, 35, 18, 32, 18,
       74, 32, 18, 32, 18, 74, 32, 18, 32, 35, 18, 49, 77, 77, 49, 77, 49,
       49])

In [66]:
ex_q_seqs

[array([67, 18, 67, 18, 67, 18, 18, 67, 67, 18, 18, 67, 32, 35, 18, 32, 18,
        74, 32, 18, 32, 18, 74, 32, 18, 32, 35, 18, 49, 77, 77, 49, 77, 49,
        49]),
 array([48, 99, 48, 99, 48, 99, 99, 48, 49, 99, 99, 49, 49, 99, 54, 54, 54,
        54, 54, 54, 54, 54, 54, 54]),
 array([31, 31, 31, 31, 30, 30]),
 array([ 4, 34, 57, 90, 26, 26, 26]),
 array([  4,   4,   4,   4,   3,   3,  54,  54,  22,  84,  54,  52,   3,
          2,   2,   2,   2,   2,   1,   1,   1,   2,   4,  54,   2,   2,
          4,  61,  62,  54,  58,  58,  58,  93,  98,  98,  98,  98,  56,
         16,  73,  99,  99,  39,  73,  73,  77,  99,  99,  77,  77,  99,
         77,  99,  51,  99,  51,  99,  23,   0,  96,   0,   1,   0,   1,
          0,  96,  62,  60,  74,  61,  31,  31,  31,  31,  31,  31,  31,
         31,  25,  20,  20,  20,  20,  20,  19,  19,  19,  19,  19,  19,
         49,  99,  49,  99,  99,  49,  16,   9,   9,   9,   9,   9,   1,
          0,   1,   0,   3,   3,   3,   3,   4,   2,   2,   1,  

In [69]:
from torch.cuda import FloatTensor

tmp_q_seqs = []
tmp_qshft_seqs = []


for q_s in ex_q_seqs:
    tmp_q_seqs.append(FloatTensor(q_s[:-1]))
    tmp_qshft_seqs.append(FloatTensor(q_s[1:]))    
    

In [70]:
tmp_q_seqs

[tensor([67., 18., 67., 18., 67., 18., 18., 67., 67., 18., 18., 67., 32., 35.,
         18., 32., 18., 74., 32., 18., 32., 18., 74., 32., 18., 32., 35., 18.,
         49., 77., 77., 49., 77., 49.]),
 tensor([48., 99., 48., 99., 48., 99., 99., 48., 49., 99., 99., 49., 49., 99.,
         54., 54., 54., 54., 54., 54., 54., 54., 54.]),
 tensor([31., 31., 31., 31., 30.]),
 tensor([ 4., 34., 57., 90., 26., 26.]),
 tensor([  4.,   4.,   4.,   4.,   3.,   3.,  54.,  54.,  22.,  84.,  54.,  52.,
           3.,   2.,   2.,   2.,   2.,   2.,   1.,   1.,   1.,   2.,   4.,  54.,
           2.,   2.,   4.,  61.,  62.,  54.,  58.,  58.,  58.,  93.,  98.,  98.,
          98.,  98.,  56.,  16.,  73.,  99.,  99.,  39.,  73.,  73.,  77.,  99.,
          99.,  77.,  77.,  99.,  77.,  99.,  51.,  99.,  51.,  99.,  23.,   0.,
          96.,   0.,   1.,   0.,   1.,   0.,  96.,  62.,  60.,  74.,  61.,  31.,
          31.,  31.,  31.,  31.,  31.,  31.,  31.,  25.,  20.,  20.,  20.,  20.,
          20.,  19.,  

In [71]:
tmp_qshft_seqs

[tensor([18., 67., 18., 67., 18., 18., 67., 67., 18., 18., 67., 32., 35., 18.,
         32., 18., 74., 32., 18., 32., 18., 74., 32., 18., 32., 35., 18., 49.,
         77., 77., 49., 77., 49., 49.]),
 tensor([99., 48., 99., 48., 99., 99., 48., 49., 99., 99., 49., 49., 99., 54.,
         54., 54., 54., 54., 54., 54., 54., 54., 54.]),
 tensor([31., 31., 31., 30., 30.]),
 tensor([34., 57., 90., 26., 26., 26.]),
 tensor([  4.,   4.,   4.,   3.,   3.,  54.,  54.,  22.,  84.,  54.,  52.,   3.,
           2.,   2.,   2.,   2.,   2.,   1.,   1.,   1.,   2.,   4.,  54.,   2.,
           2.,   4.,  61.,  62.,  54.,  58.,  58.,  58.,  93.,  98.,  98.,  98.,
          98.,  56.,  16.,  73.,  99.,  99.,  39.,  73.,  73.,  77.,  99.,  99.,
          77.,  77.,  99.,  77.,  99.,  51.,  99.,  51.,  99.,  23.,   0.,  96.,
           0.,   1.,   0.,   1.,   0.,  96.,  62.,  60.,  74.,  61.,  31.,  31.,
          31.,  31.,  31.,  31.,  31.,  31.,  25.,  20.,  20.,  20.,  20.,  20.,
          19.,  19.,  

In [72]:
from torch.nn.utils.rnn import pad_sequence

# from torch.cuda import FloatTensor
# print(len(FloatTensor(q_seq[:-1])))
# FloatTensor(q_seq[:-1])

tmp_q_seqs = pad_sequence(tmp_q_seqs, batch_first=True, padding_value=-1)
tmp_qshft_seqs = pad_sequence(tmp_qshft_seqs, batch_first=True, padding_value=-1)


In [77]:
mask_seqs = (tmp_q_seqs != -1) * (tmp_qshft_seqs != -1)

In [ ]:
tmp_q_seqs = tmp_q_seqs * mask_seqs


In [81]:
tmp_qshft_seqs = tmp_qshft_seqs * mask_seqs


In [82]:
# list(tmp_qshft_seqs[0])

[tensor(18.),
 tensor(67.),
 tensor(18.),
 tensor(67.),
 tensor(18.),
 tensor(18.),
 tensor(67.),
 tensor(67.),
 tensor(18.),
 tensor(18.),
 tensor(67.),
 tensor(32.),
 tensor(35.),
 tensor(18.),
 tensor(32.),
 tensor(18.),
 tensor(74.),
 tensor(32.),
 tensor(18.),
 tensor(32.),
 tensor(18.),
 tensor(74.),
 tensor(32.),
 tensor(18.),
 tensor(32.),
 tensor(35.),
 tensor(18.),
 tensor(49.),
 tensor(77.),
 tensor(77.),
 tensor(49.),
 tensor(77.),
 tensor(49.),
 tensor(49.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tensor(-0.),
 tenso

**training 부분 살펴보기**

In [104]:
(model(q.long(), r.long()) * one_hot(qshft.long(), 110)).sum(-1).shape #q, r = q_seq, r_seq 마지막거 빼고

torch.Size([621, 100])

In [81]:
rshft

tensor([[0., 1., 1.,  ..., 1., 1., 1.],
        [1., 0., 0.,  ..., -0., -0., -0.],
        [1., 1., 1.,  ..., 1., 0., 1.],
        ...,
        [0., 0., -0.,  ..., -0., -0., -0.],
        [1., 1., 0.,  ..., -0., -0., -0.],
        [0., 0., 1.,  ..., 0., 1., 1.]])

In [173]:
m

tensor([[ True,  True,  True,  ..., False, False, False],
        [ True,  True,  True,  ...,  True,  True,  True],
        [ True,  True,  True,  ..., False, False, False],
        ...,
        [ True,  True,  True,  ..., False, False, False],
        [ True,  True,  True,  ..., False, False, False],
        [ True,  True,  True,  ...,  True,  True,  True]])

In [174]:
q

tensor([[ 30.,  30.,  30.,  ...,  -0.,  -0.,  -0.],
        [ 59.,  59.,  59.,  ...,  32.,  35.,  72.],
        [  2.,   2.,   2.,  ...,  -0.,  -0.,  -0.],
        ...,
        [ 23.,  23.,  23.,  ...,  -0.,  -0.,  -0.],
        [107.,  34.,  34.,  ...,  -0.,  -0.,  -0.],
        [  1.,   1.,   1.,  ...,  23.,  23.,  23.]])

In [175]:
# forward 에 해당(모델 archi)

q.long()+110*r.long()

tensor([[140,  30, 140,  ...,   0,   0,   0],
        [ 59, 169,  59,  ..., 142, 145,  72],
        [  2, 112, 112,  ...,   0,   0,   0],
        ...,
        [133, 133, 133,  ...,   0,   0,   0],
        [217,  34, 144,  ...,   0,   0,   0],
        [  1, 111, 111,  ...,  23, 133, 133]])

In [176]:
# forward 에 해당(모델 archi)

from torch.nn import Module, Embedding, LSTM, Linear, Dropout
a = Embedding(110 * 2, 100)
a(q.long()+110*r.long())

tensor([[[ 0.0646,  0.6782,  0.4805,  ...,  0.3244,  1.4673, -0.8872],
         [ 0.3542, -1.1548,  0.2123,  ...,  1.3143, -0.7638, -0.7568],
         [ 0.0646,  0.6782,  0.4805,  ...,  0.3244,  1.4673, -0.8872],
         ...,
         [-1.1790,  0.5804,  0.6915,  ...,  0.9621, -2.3035,  0.6378],
         [-1.1790,  0.5804,  0.6915,  ...,  0.9621, -2.3035,  0.6378],
         [-1.1790,  0.5804,  0.6915,  ...,  0.9621, -2.3035,  0.6378]],

        [[ 0.7912,  0.7721, -2.0852,  ..., -0.5702, -1.1564, -0.3657],
         [-0.4141,  0.4807,  0.4325,  ...,  1.3115, -0.4596,  0.0527],
         [ 0.7912,  0.7721, -2.0852,  ..., -0.5702, -1.1564, -0.3657],
         ...,
         [-0.0816,  0.3652,  0.4768,  ..., -1.3596,  1.9532, -0.3649],
         [ 0.3551,  0.6378, -1.2782,  ...,  0.2093, -1.3833, -0.7918],
         [ 0.0582,  0.5138,  1.4762,  ...,  0.6449, -0.0457,  0.8245]],

        [[-0.6972, -1.1253,  0.0500,  ...,  1.1588, -0.0716, -0.8260],
         [ 0.3127,  1.5683, -0.3202,  ..., -0

In [177]:
# forward 에 해당(모델 archi)

b = LSTM(100, 100, batch_first=True)
c = Linear(100, 110)
d = Dropout()


In [186]:
# forward 에 해당(모델 archi)

e, _ = b(a(q.long()+110*r.long()))
torch.sigmoid(d(c(e))) # forward 결과 y

tensor([[[0.5000, 0.5000, 0.5000,  ..., 0.5246, 0.4426, 0.4735],
         [0.5000, 0.5000, 0.5626,  ..., 0.5378, 0.4564, 0.5000],
         [0.5000, 0.4902, 0.5018,  ..., 0.5226, 0.5000, 0.5000],
         ...,
         [0.5418, 0.5000, 0.5000,  ..., 0.4750, 0.4946, 0.6006],
         [0.5460, 0.4337, 0.5000,  ..., 0.5000, 0.4912, 0.6016],
         [0.5000, 0.5000, 0.5128,  ..., 0.4721, 0.4900, 0.6014]],

        [[0.5000, 0.5000, 0.5476,  ..., 0.5000, 0.5000, 0.5000],
         [0.5560, 0.4343, 0.5379,  ..., 0.4528, 0.5215, 0.5889],
         [0.5968, 0.5000, 0.5486,  ..., 0.4497, 0.5000, 0.5364],
         ...,
         [0.5000, 0.5000, 0.5113,  ..., 0.5000, 0.4716, 0.5527],
         [0.5000, 0.4449, 0.5000,  ..., 0.5000, 0.5000, 0.5000],
         [0.4397, 0.5000, 0.4153,  ..., 0.4800, 0.5000, 0.5000]],

        [[0.5000, 0.4369, 0.4506,  ..., 0.4888, 0.4753, 0.4686],
         [0.5000, 0.4776, 0.4750,  ..., 0.5000, 0.5431, 0.5161],
         [0.5000, 0.5000, 0.5000,  ..., 0.4162, 0.5598, 0.

## 이거랑(아래)

In [187]:
model(q.long(), r.long())#.shape #[621, 100, 110] #embedding+LSTM

tensor([[[0.5180, 0.5133, 0.5491,  ..., 0.4413, 0.5101, 0.5148],
         [0.4400, 0.4519, 0.5395,  ..., 0.4244, 0.4993, 0.4693],
         [0.4943, 0.4993, 0.5656,  ..., 0.4047, 0.4955, 0.4871],
         ...,
         [0.5774, 0.0921, 0.5310,  ..., 0.3919, 0.4639, 0.2780],
         [0.5343, 0.0875, 0.5200,  ..., 0.4038, 0.4630, 0.2393],
         [0.5116, 0.0902, 0.5057,  ..., 0.4063, 0.4607, 0.2148]],

        [[0.3919, 0.5343, 0.4814,  ..., 0.5000, 0.4836, 0.4765],
         [0.5019, 0.5833, 0.5576,  ..., 0.4024, 0.4867, 0.5294],
         [0.4584, 0.5923, 0.5009,  ..., 0.4640, 0.4882, 0.5255],
         ...,
         [0.7182, 0.5946, 0.5686,  ..., 0.5039, 0.5107, 0.4482],
         [0.8071, 0.5494, 0.6340,  ..., 0.4277, 0.5068, 0.6270],
         [0.7743, 0.5369, 0.5571,  ..., 0.3988, 0.3290, 0.2960]],

        [[0.4820, 0.5091, 0.4890,  ..., 0.4439, 0.4985, 0.5165],
         [0.6148, 0.6036, 0.6547,  ..., 0.4435, 0.5187, 0.6615],
         [0.6284, 0.6215, 0.6781,  ..., 0.4446, 0.5184, 0.

In [191]:
qshft.long()[0] #[621,100] #여기서 숫자(스킬) 하나가 아래에서 한줄(110열 원핫인코딩에 해당)

tensor([30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30,
        30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 31, 23, 23, 23, 23, 23, 23, 23,
        31, 31, 31, 31, 31, 31, 31, 31, 31, 57, 90, 57, 90, 90, 57, 90, 57, 57,
        90, 90, 57, 90, 57, 57, 90, 90, 57, 57, 90, 95, 77, 95, 77, 95, 51, 73,
        73,  1, 99, 99,  1, 99,  1,  1, 99,  1, 99,  1, 99,  1, 99, 77, 48, 99,
        48, 99, 72, 72, 72,  0,  0,  0,  0,  0])

## 이거랑(아래) 곱함

In [87]:
# 각 데이터셋에 대해 100차원으로 임베딩하는데, skill 110개 -> one_hot encoding
# 데이터셋 1개 당(한 학생의 풀이로그) [100,110] 크기로 인코딩
# 왜 100차원이냐면, seq_len 을 100으로 맞췄으니까 match_seq_len 을 통해서!!!

# 그래서 아래는 스킬셋에 대한 정보 임베딩 값을 가지는 거임


one_hot(qshft.long(), 110)#[0][3]#.shape #[621, 100, 110]

tensor([[[0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         ...,
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0]],

        [[0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         ...,
         [1, 0, 0,  ..., 0, 0, 0],
         [1, 0, 0,  ..., 0, 0, 0],
         [1, 0, 0,  ..., 0, 0, 0]],

        [[0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         ...,
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0]],

        ...,

        [[0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [1, 0, 0,  ..., 0, 0, 0],
         ...,
         [1, 0, 0,  ..., 0, 0, 0],
         [1, 0, 0,  ..., 0, 0, 0],
         [1, 0, 0,  ..., 0, 0, 0]],

        [[0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0,

In [193]:
one_hot(qshft.long(), 110)[0][3]

tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [203]:
(model(q.long(), r.long()) * one_hot(qshft.long(), 110)).sum(-1)#.shape [621,100]

# 이렇게 하면 다음 Time Step 에서 푼 문제에 대해서만 Backpropagation 을 수행하게 되고,
# 결과적으로 다음 Time Step 에서 푼 문제만을 고려하여 학습이 수행되게 됨

tensor([[0.5875, 0.6822, 0.6508,  ..., 0.5774, 0.5343, 0.5116],
        [0.4487, 0.5645, 0.5042,  ..., 0.7604, 0.4737, 0.5245],
        [0.4890, 0.6547, 0.6781,  ..., 0.3871, 0.3871, 0.3871],
        ...,
        [0.5915, 0.6223, 0.6326,  ..., 0.3871, 0.3871, 0.3871],
        [0.5517, 0.5791, 0.6710,  ..., 0.3871, 0.3871, 0.3871],
        [0.5211, 0.6403, 0.6556,  ..., 0.5835, 0.6008, 0.6110]],
       grad_fn=<SumBackward1>)

In [89]:
dataset.__getitem__(0)

(array([67, 18, 67, 18, 67, 18, 18, 67, 67, 18, 18, 67, 32, 35, 18, 32, 18,
        74, 32, 18, 32, 18, 74, 32, 18, 32, 35, 18, 49, 77, 77, 49, 77, 49,
        49, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]),
 array([ 0,  0,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  1,  1,  0,  0,  0,  1,  1,  0,  0,  0,  0,  1,  1,  0,  1,  0,
         0, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]))

In [90]:
train_dataset.__getitem__(0)

(array([109, 109, 109, 109, 107, 107, 107, 107, 107, 107, 107, 107, 107,
        107, 107, 107, 107, 107, 107, 107, 107, 107, 107, 107, 102,   9,
          9,   9,   9,   9,   9,   9,   9,  13,  10,  10,  10,  10,  10,
         10,  10,  11,   9,   9,   9,   9,   9, 102, 102,  65,  96, 107,
        107, 107, 107, 107, 107,  19,  19,  19,  19,  19,  19,  19,  19,
         19,  19,  19,  19,  19,  19,   4, 101,  69,  16,  16,  16,  16,
         16,  16,  16,  16,  16,  16,  16,  16,  16,  16,  16,  16,  16,
         16,  16,  16,  16,  16,  16,  16,  16,  29,  29]),
 array([1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0,
        1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1,
        1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1,
        0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0]))

In [95]:
test_dataset.__getitem__(0)

(array([ 1,  1,  1,  1,  1,  3,  3,  3,  3,  3,  3, 54, 54, 54, 54, 54, 54,
        54, 54, 54, 54, 54, 54, 54, 94, 94,  4,  4,  4,  4, 55, 55, 55, 23,
        23, 23, 23, 23, 23, 23, 23, 23, 23, 52, 52, 52, 52, 52, 52, 28, 28,
        28, 28, 28, 28, 28, 28, 28, 28,  2,  2,  2,  2,  2,  2,  2,  2,  2,
         2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  4,  4,
        28, 28, 28, 28,  2,  2,  2,  2,  2,  2, 52, 52, 52,  2,  2,  2]),
 array([1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0,
        0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1]))